[FSDP] Removed clamp to `NO_SHARD` for world size 1

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120334
* #124764

This is a workaround for https://github.com/pytorch/pytorch/issues/119937. `NO_SHARD` always using unsharded views breaks with `DTensor` extensions since some logic assumes that sharded views are `torch.Tensors`, but `NO_SHARD` makes sharded views and unsharded views the same (as `DTensor`s).

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @rohan-varma

Commit:
[FSDP] Removed clamp to `NO_SHARD` for world size 1

[ghstack-poisoned]

Commit:
Update on "[FSDP] Removed clamp to `NO_SHARD` for world size 1"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "[FSDP] Removed clamp to `NO_SHARD` for world size 1"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

mvpatel2000 commented:
@awgu maybe it's worth emitting a warning still that NO_SHARD might be more performant in this case? Not strong opinion though

awgu commented:
We have been seeing some issues with `NO_SHARD` internally, so we were planning to deprecate it. We do want to provide a good solution for a replicated data parallelism that can compose with FSDP (and we have started discussing this).

mvpatel2000 commented:
https://github.com/pytorch/pytorch/issues/119937#issuecomment-1964671752

Flagging that I still see SHARD_GRAD_OP issue with this :( (but FULL_SHARD works fine!)

Commit:
Update on "[FSDP] Removed clamp to `NO_SHARD` for world size 1"


This is a workaround for https://github.com/pytorch/pytorch/issues/119937. `NO_SHARD` always using unsharded views breaks with `DTensor` extensions since some logic assumes that sharded views are `torch.Tensors`, but `NO_SHARD` makes sharded views and unsharded views the same (as `DTensor`s).

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "[FSDP] Removed clamp to `NO_SHARD` for world size 1"


This is a workaround for https://github.com/pytorch/pytorch/issues/119937. `NO_SHARD` always using unsharded views breaks with `DTensor` extensions since some logic assumes that sharded views are `torch.Tensors`, but `NO_SHARD` makes sharded views and unsharded views the same (as `DTensor`s).

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "[FSDP] Removed clamp to `NO_SHARD` for world size 1"


This is a workaround for https://github.com/pytorch/pytorch/issues/119937. `NO_SHARD` always using unsharded views breaks with `DTensor` extensions since some logic assumes that sharded views are `torch.Tensors`, but `NO_SHARD` makes sharded views and unsharded views the same (as `DTensor`s).

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "[FSDP] Removed clamp to `NO_SHARD` for world size 1"


This is a workaround for https://github.com/pytorch/pytorch/issues/119937. `NO_SHARD` always using unsharded views breaks with `DTensor` extensions since some logic assumes that sharded views are `torch.Tensors`, but `NO_SHARD` makes sharded views and unsharded views the same (as `DTensor`s).

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "[FSDP] Removed clamp to `NO_SHARD` for world size 1"


This is a workaround for https://github.com/pytorch/pytorch/issues/119937. `NO_SHARD` always using unsharded views breaks with `DTensor` extensions since some logic assumes that sharded views are `torch.Tensors`, but `NO_SHARD` makes sharded views and unsharded views the same (as `DTensor`s).

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "[FSDP] Removed clamp to `NO_SHARD` for world size 1"


This is a workaround for https://github.com/pytorch/pytorch/issues/119937. `NO_SHARD` always using unsharded views breaks with `DTensor` extensions since some logic assumes that sharded views are `torch.Tensors`, but `NO_SHARD` makes sharded views and unsharded views the same (as `DTensor`s).

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]
